pose_evaluation/metrics/embedding_distance_metric.py

```diff
@@ -0,0 +1,104 @@
+        """
+        super().__init__(f"EmbeddingDistanceMetric {kind}", higher_is_better=False)
+        self.kind = kind
+        if device is None:
+            self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        else:
+            self.device = torch.device(device) if isinstance(device, str) else device
+
+    def _to_tensor(self, data: Union[np.ndarray, torch.Tensor]) -> torch.Tensor:
+        """
+        Convert input to a PyTorch tensor if it is a NumPy array.
+
+        Args:
+            data (np.ndarray | torch.Tensor): Input data.
+
+        Returns:
+            torch.Tensor: Tensor on the correct device.
+        """
+        if isinstance(data, np.ndarray):
+            data = torch.tensor(data, dtype=torch.float32)
+        return data.to(self.device)
+
+    def score(self, hypothesis: Union[np.ndarray, torch.Tensor], reference: Union[np.ndarray, torch.Tensor]) -> float:
+        """
+        Compute the distance between two embeddings.
+
+        Args:
+            hypothesis (np.ndarray | torch.Tensor): A single embedding vector.
+            reference (np.ndarray | torch.Tensor): Another single embedding vector.
+
+        Returns:
+            float: The calculated distance.
+        """
+        hypothesis = self._to_tensor(hypothesis)
+        reference = self._to_tensor(reference)
+
+        if self.kind == "cosine":
+            # Normalize both embeddings to unit length
+            hypothesis = F.normalize(hypothesis, p=2, dim=0)
+            reference = F.normalize(reference, p=2, dim=0)
+            # Cosine similarity, converted to distance
+            similarity = torch.dot(hypothesis, reference).item()
+            return 1 - similarity
+        if self.kind == "l2":
+            # L2 distance
+            return torch.norm(hypothesis - reference).item()
+
+        raise ValueError(f"Unsupported distance metric: {self.kind}")
+
+    def score_all(
+        self,
+        hypotheses: List[Union[np.ndarray, torch.Tensor]],
+        references: List[Union[np.ndarray, torch.Tensor]],
+        progress_bar: bool = True,
+    ) -> torch.Tensor:
+        """
+        Compute the pairwise distance between all hypotheses and references. Expects 2D inputs.
+
+        Args:
+            hypotheses (list[np.ndarray | torch.Tensor]): List of hypothesis embeddings.
+            references (list[np.ndarray | torch.Tensor]): List of reference embeddings.
+            progress_bar (bool): Whether to display a progress bar.
+
+        Returns:
+            torch.Tensor, distance matrix. Row i is the distances of hypotheses[i] to all rows of references
+        """
+        # Convert inputs to tensors and stack
+        hypotheses = torch.stack([self._to_tensor(h) for h in hypotheses])
+        references = torch.stack([self._to_tensor(r) for r in references])
+
+        if self.kind == "cosine":
+            # Normalize the tensors along the feature dimension (dim=1)
```

Comment:
this is a reimplementation of `signclip_distance...` see my comments there, they are relevant here too

Reply:
Right. I (1) made signclip distance, (2) realized nothing in there was signclip-specific, and (3) made this one instead. So I was planning to either take signclip distance  out or just subclass

Reply:
then i think, for now, let's "remove" signclip_distance, and focus on embedding metric. signclip is just an implementation, that can extend it as you say (takes poses converts to vectors first)

Reply:
Sounds great to me!